plugins/qCompass/ccTrace.h

```diff
@@ -0,0 +1,199 @@
+#ifndef CC_TRACE_HEADER
+#define CC_TRACE_HEADER
+
+//#define DEBUG_PATH   //n.b. uncomment this to write point-costs to a scalar field for debug purposes
+
+#include <ccHObject.h>
+#include <ccPolyline.h>
+#include <ccSphere.h>
+#include <DgmOctree.h>
+#include <DgmOctreeReferenceCloud.h>
+#include <GenericIndexedCloudPersist.h>
+#include <ccPointCloud.h>
+#include <ccColorTypes.h>
+#include <Neighbourhood.h>
+#include <ccPlane.h>
+#include <Jacobi.h>
+#include <ccScalarField.h>
+
+#include <vector>
+#include <algorithm>
+#include <unordered_map>
+#include <deque>
+
+/*
+A ccTrace object is essentially a ccPolyline that is controlled/created by "waypoints" and a least-cost path algorithm
+designe to pick fracture traces and lithological contacts (i.e. follow edges or linear features in RGB space). 
+
+If treated as a ccPolyline, it will behave as a conventional polyline going through a series of points (including intermediate
+points between the waypoints). 
+
+If treated as a ccTrace object, then the waypoints can be manipulated and the underlying ccPolyline recalculated. For convenience, 
+the waypoints are also drawn as bubbles.
+
+*/
+class ccTrace : public ccPolyline
+{
+public:
+	ccTrace(ccPointCloud* associatedCloud);
+	virtual ~ccTrace() {}
+
+	//inherited from ccHObject
+	inline virtual CC_CLASS_ENUM getClassID() const override { return CC_TYPES::POLY_LINE; }
+
+	/*
+	Adds waypoint to the end of this trace.
+	@Args
+	 *pointIndex* = the index of the point (in the cloud associated with this object) that is the new waypoint
+	*/
+	void pushWaypoint(int pointId) { m_waypoints.push_back(pointId); }//addPointIndex(id);}
+
+	/*
+	Deletes the specified waypoint.
+	@Args
+	*pointIndex* = the index of the point (in the cloud associated with this object) that represents the waypoint to be deleted. If this point is not
+	               a waypoint then the function does nothing.
+	*/
+	void deleteWaypoint(int pointId) { 
+		m_waypoints.erase(std::remove(m_waypoints.begin(), m_waypoints.end(), pointId), m_waypoints.end()); 
+	}
+
+	/*
+	Tries to insert/append the given waypoint based on its location. If the point is within a "circle" including the start and end of the closest segment
+	then it "inserts" the point. Otherwise the point is appended onto the end of the trace.
+
+	Returns True if the point new waypoint was succesfully added/inserted and the path succesfully updated.
+	*/
+	int insertWaypoint(int pointId);
+
+	/*
+	Undo last action
+	*/
+	void undoLast()
+	{
+		if (m_previous != -1)
+			m_waypoints.erase(m_waypoints.begin() + m_previous);
```

Comment:
This is exactly the reason to always use {}. The code says only this line is executed for the `if`, but the indentation seems to imply that the `m_trace` line should be too...